### PR TITLE
#19854 Use SSHJ tunneling for jump server support

### LINKED ISSUE #19854: SSH Tunnel: Incorrect hostname `localhost` is checked against the SSH host certificate
### Description

The `SSHJ` SSH tunnel implementation incorrectly checks `localhost` against the SSH server host certificate.

This is particularly important for StepCA or Teleport users, where the host certs likely contain the servers IP, and **not** `localhost`.

Note that this error only occurs when using a jump server. The `Local host`, `Remote host`, and `Bypass host verification` options don't have any impact.

### DBeaver Version

Community Edition 23.0.3.202304240616

### Operating System

Windows 11

### Database and driver

Not applicable - the error happens _before_ connecting to the database.

The issue is present for at least these two drivers:
- `mysql:mysql-connector-java` 8.0.29
- `org.postgresql:postgresql` 42.5.0

### Steps to reproduce

You will need:
1. An SSH jump server
2. A remote database server
   - Must be accessible via SSH
   - With _or_ without the jump server
4. Your `known_hosts` configured to accept hostkeys from both the database & jump servers
5. The host certificates for both servers are signed 

Then, in DBeaver:
1. Create a new connection. Any driver will work (I tested with PostgreSQL and MySQL).
2. Enable SSH Tunneling
   - Pick `SSHJ` as the implementation
   - All other advanced settings left as default
3. Ensure the `Use jump server` section is enabled & filled out
4. Click `Test tunnel configuration`

An error is thrown: `KeyExchange certificate check failed: Hostname 'localhost' doesn't match any of the principals: 'xxx.xxx.xxx.xxx'`.

Then, try:
1. Disable `Use jump server`
2. Click `Test tunnel configuration`

The SSH connection works without issues.

### Additional context

[dbeaver.log](https://github.com/dbeaver/dbeaver/files/11377346/dbeaver.log)

### PR DESCRIPTION
**Pull request by ryanc-me**
Consider the following setup:
- An SSH `jump-server` at `jump.example.com:22`
- A main `ssh-server` at `10.0.0.1:22`, requires jumping through `jump-server`
- The SSH host certificates are signed with `jump.example.com` and `10.0.0.1` respectively

Currently DBeaver will:
- Connect to the `jump-server`
- Set up a port forward (tunnel) from `localhost:12345` -> `10.0.0.1:22` via `jump-server`
- Connect to `localhost:12345`

However, the SSH host certificate check fails because `localhost` is not in the cert (only `10.0.0.1` is).

The fix is to follow DBeaver's [`Jump.java` example](https://github.com/hierynomus/sshj/blob/9e8bef24c5dcf5353677333037d5a52ac3f3a34f/examples/src/main/java/net/schmizz/sshj/examples/Jump.java), where they suggest using [`jumpServer.newDirectConnection()`](https://github.com/hierynomus/sshj/blob/9e8bef24c5dcf5353677333037d5a52ac3f3a34f/src/main/java/net/schmizz/sshj/SSHClient.java#L676-L688).

This method will correctly check `10.0.0.1` against the host certificate. 

<br />

See #19854 for more context.

### CODE REVIEW

**Comment by ryanc-me** [comment]
Rebased on `dbeaver/devel`

**Review by ShadelessFox — APPROVED** [review]

File: plugins/org.jkiss.dbeaver.net.ssh.sshj/src/org/jkiss/dbeaver/model/net/ssh/SSHImplementationSshj.java
Comment (E1izabeth) [inline plugins/org.jkiss.dbeaver.net.ssh.sshj/src/org/jkiss/dbeaver/model/net/ssh/SSHImplementationSshj.java:90]: `setPortForwarding` method was used here, but it contains not only socket creation. We miss `LocalPortListener` by replacing this with `newDirectConnection` method. Don't we need it?
``` 
setPortForwarding(...) {
        ...
        final LocalPortForwarder forwarder = client.newLocalPortForwarder(parameters, ss);
        final LocalPortListener listener = new LocalPortListener(forwarder, parameters);
        listener.start();
        listeners.add(listener);
        ...
}
```

Comment (ShadelessFox) [reply plugins/org.jkiss.dbeaver.net.ssh.sshj/src/org/jkiss/dbeaver/model/net/ssh/SSHImplementationSshj.java:90]: It looks like it's not used anymore and, therefore, can be removed.

Comment (E1izabeth) [reply plugins/org.jkiss.dbeaver.net.ssh.sshj/src/org/jkiss/dbeaver/model/net/ssh/SSHImplementationSshj.java:90]: Okay

**Review by E1izabeth — APPROVED** [review]

**Comment by ryanc-me** [comment]
Thanks @E1izabeth and @ShadelessFox!

Is there anything else you need me to do before merge?

**Comment by E1izabeth** [comment]
Your PR cannot be built. See the log part
```
[ERROR] Failed to execute goal org.eclipse.tycho:tycho-compiler-plugin:3.0.1:compile (default-compile) on project org.jkiss.dbeaver.net.ssh.sshj: Compilation failure: Compilation failure: 
[ERROR] /var/jenkins_home/jobs/DBeaverDesktop/jobs/dbeaver-ce-github-checks/branches/PR-19857/workspace/dbeaver/plugins/org.jkiss.dbeaver.net.ssh.sshj/src/org/jkiss/dbeaver/model/net/ssh/SSHImplementationSshj.java:[95] 
[ERROR] 	prevClient.getHostname(), prevClient.getPort()));
[ERROR] 	           ^^^^^^^^^^^
[ERROR] The method getHostname() is undefined for the type SSHClient
[ERROR] /var/jenkins_home/jobs/DBeaverDesktop/jobs/dbeaver-ce-github-checks/branches/PR-19857/workspace/dbeaver/plugins/org.jkiss.dbeaver.net.ssh.sshj/src/org/jkiss/dbeaver/model/net/ssh/SSHImplementationSshj.java:[95] 
[ERROR] 	prevClient.getHostname(), prevClient.getPort()));
[ERROR] 	                                     ^^^^^^^
[ERROR] The method getPort() is undefined for the type SSHClient
[ERROR] 2 problems (2 errors)
[ERROR] -> [Help 1]
```

Which version of sshj do you use locally?

**Comment by ryanc-me** [comment]
@E1izabeth Good catch - not sure why that didn't throw an error during my testing.

I think we should be using [`prevClient.getRemoteHostname()`](https://github.com/hierynomus/sshj/blob/master/src/main/java/net/schmizz/sshj/SocketClient.java#L147-L149) and [`prevClient.getRemotePort()`](https://github.com/hierynomus/sshj/blob/master/src/main/java/net/schmizz/sshj/SocketClient.java#L151-L153).

Would you like me to make a new commit?

**Comment by E1izabeth** [comment]
Yeah, sure, you can make new commits. I hope it'll help to build the solution.
Thank you for your time.

**Review by E1izabeth — APPROVED** [review]

**Review by E1izabeth — APPROVED** [review]

**Comment by ryanc-me** [comment]
@E1izabeth just touching base - what is the policy around merging PRs? I have a few staff who are blocked by this issue, it'd be great to give them some indication when this might be fixed

**Comment by ShadelessFox** [comment]
Hello @ryanc-me,

It's awaiting our QA team. Unfortunately, it didn't make it into the 23.0.5 release but will likely reach the 23.1.0.

**Comment by E1izabeth** [comment]
At first, developers review your PR. When all comments are resolved, PR goes to our QA team. If no bugs found, it is successfully merged. If tester find something that doesn't not work with your changes, you receive a comment about that. Then we wait, while you fix it and help you if needed. And after that, we come to the beginning and each step repeats.

Currently both of your pull requests are waiting for the QA team.

**Comment by E1izabeth** [comment]
At first, developers reviews your PR. When all comments are resolved, PR goes to our QA team. If no bugs found, it will be successfully merged. If tester will find something that will not work with your changes, you will receive a comment about that and we will wait, while you  fix it and help you if needed. Then we come to the beginning and each step repeats.

Currently both of your pull requests are waiting for the QA team.

**Comment by Matvey16** [comment]
Verified

**Pull request merged**
